src/ImageSharp.Web/Caching/CacheKey.cs

```diff
@@ -0,0 +1,37 @@
+// Copyright (c) Six Labors.
+// Licensed under the Apache License, Version 2.0.
+
+using System.Text;
+using Microsoft.AspNetCore.Http;
+using SixLabors.ImageSharp.Web.Commands;
+
+namespace SixLabors.ImageSharp.Web.Caching
+{
+    /// <summary>
+    /// Creates a cache key based on the lowercased request host, path and commands.
+    /// </summary>
+    public class CacheKey : ICacheKey
```

Comment:
I think we should only ship with the two implementations as described in https://github.com/SixLabors/ImageSharp.Web/pull/147#issuecomment-1023047344 although I like your naming convention better.

- `UriAbsoluteCacheKey`
- `UriRelativeCacheKey`

I can't see a scenario where we would want a case sensitive cache key e.g `UriAbsoluteCacheKey` as that opens us up to potential DOS so we should always convert the string to lower invariant.

From a performance perspective we should use `string.Create` in conjunction with `Span<char>.ToLowerInvarient()` to ensure that string creation is snappy. [`Urihelper`](https://source.dot.net/#Microsoft.AspNetCore.Http.Extensions/UriHelper.cs,e32d072f616dc38d) has optimized code you can use as a reference but we'd have to change a few things to do the required casing work.

Reply:
Depending on the OS (or file system configuration: https://docs.microsoft.com/en-us/windows/wsl/case-sensitivity#modify-case-sensitivity), the path can be case sensitive, which means different casing can resolve to different files and also that wrong casing wouldn't find the original image in the first place... So the cache hash should only be 'normalized' when the file system is case insensitive. I'm not sure if this can be easily determined at runtime, so adding a setting for this would be sensible.

The parsed commands are already case insensitive (or at least the keys), because the default implementation uses an `InvariantIgnoreCase` comparer (and gets them from the query string, which isn't case sensitive according to the RFCs). Although it ultimately depends on the processor, command values should also be case insensitive (enum parsing already ignores casing). It would be good to ensure this value is always lowercased, so changes in the command casing won't cause different cache hashes.

Reply:
Case sensitivity of file requests is dealt with via individual  provider implementations. They don’t use the key to return the source image, they use the input uri. 

Only the cache uses the key and it’s imperative that the generated hash is case insensitive.

Reply:
Yep. Case sensitivity of file requests is dealt with via individual  provider implementations. They don’t use the key to return the source image, they use the input uri. If nothing is found, nothing is cached.

Rethinking about this then you're correct but the implementations should be non inheriting since we don't want a virtual method as a default implementation on a fast path. We're already paying a virtualization cost via interfaces.

So, as you say, to cover our backs, your implementations would be:

- `UriAbsoluteCacheKey`
- `UriRelativeCacheKey`
- `UriAbsoluteCaseInsensitiveCacheKey`
- `UriRelativeCaseInsensitiveCacheKey`

You're gonna want custom implementations of the `UriHelper` methods for handling the concatenation + casing to avoid duplicate string generation for all scenarios since the caching of commands should, as you've rightly identified, be case insensitive. A processor not using the case insensitive `CommandCollection` for command key matching would be considered a bug in that implementation.

Reply:
That's exactly my point: the `IImageProvider` implementations correctly deal with the case sensitivity, but the cache key doesn't. So if you have multiple files that only differ in casing, it will only cache 1 generated image (with the metadata next to it), depending on which was requested first, the TLRU cache, file write times and other weird factors:

![Different cased files](https://user-images.githubusercontent.com/1051287/151451475-c4fc163c-386e-4f2a-a641-bd8e3166271d.png)

You can test this yourself by running the sample site in WSL and creating the copies using `cp imagesharp-logo.png ImageSharp-logo.png`.

Having a setting to disable lowercasing the cache key would at least make it easy to switch (if you're running into this edge case), but also ensure custom implementations of `ICacheKey` are exposed to this setting if it's passed into the `Create` method. And it requires less classes/implementations 😄

Reply:
I wouldn't add an option for this for two reasons. 

1. Each implementation would have to respect that boolean option (complexity).
2. Each implementation would have to parse that boolean per request (performance). 

Making the case sensitivity a detail of the registered implementation provides us with the most flexibility. Registering an `ICacheKey` implementation is no more difficult than setting an option.

Reply:
The whole point for adding the `bool caseSensitive` parameter is to make sure the implementations _have_ to respect that setting (or at least are directly exposed to it).

The same can be said for the `uint length` parameter currently on `ICacheHash.Create`: all implementations need to ensure they generate a hash of the specified length. Not having a global setting might result in one implementation using a setting to be able to configure it and another always returning a fixed length hash (or just as worse: every implementation using their own setting).

Another very important thing about the `CacheKeyCaseSensitive` setting is that the need for enabling it depends on the OS/platform you're running you application on. You don't want to edit code/create separate builds to make your application work correctly for as case sensitive and insensitive file system 😖

Reply:
If the setting is really a no-go, I'd suggest to only provide the case insensitive relative (default) and absolute implementations.

Reply:
I can smash out the two other implementations for you if you like. They wouldn’t take me long

Reply:
I'd very much like that, as I'm not yet familiar how to best work with `string.Create` and `Span<char>` 😉

I did create some additional tests and benchmarks, so it should be easy for you to validate the implementations 🎉

Reply:
I've added the previous cache key creation code as baseline benchmark and the `UriAbsoluteCacheKey` implementation is still faster and allocates less (even though that also includes the request scheme, thus creating a longer cache key):

https://github.com/SixLabors/ImageSharp.Web/blob/8c966ad5c40368b3c561bae7e2783283b37f1a94/tests/ImageSharp.Web.Benchmarks/Caching/CacheKeyBenchmarks.cs#L35-L46

I'm sure you're able to make it even more performant by using `string.Create`/`Span<char>` directly, but the `UriHelper` methods are at least already outperforming the previous `StringBuilder` implementation 👍🏻